federatedscope/core/monitor.py

```diff
@@ -1,52 +1,168 @@
+            if not role.lower().startswith('server') or form == 'raw':
+                round_formatted_results_raw['Results_raw'] = new_results
+            elif form not in Monitor.SUPPORTED_FORMS:
+                continue
+            else:
+                for key in results.keys():
+                    dataset_name = key.split("_")[0]
+                    if f'{dataset_name}_total' not in results:
+                        raise ValueError(
+                            "Results to be formatted should be include the dataset_num in the dict,"
+                            f"with key = {dataset_name}_total")
+                    else:
+                        dataset_num = np.array(
+                            results[f'{dataset_name}_total'])
+                        if key in [
+                                f'{dataset_name}_total',
+                                f'{dataset_name}_correct'
+                        ]:
+                            new_results[key] = np.mean(new_results[key])
+
+                    if key in [
+                            f'{dataset_name}_total', f'{dataset_name}_correct'
+                    ]:
+                        new_results[key] = np.mean(new_results[key])
+                    else:
+                        all_res = np.array(copy.copy(results[key]))
+                        if form == 'weighted_avg':
+                            new_results[key] = np.sum(
+                                np.array(new_results[key]) *
+                                dataset_num) / np.sum(dataset_num)
+                        if form == "avg":
+                            new_results[key] = np.mean(new_results[key])
+                        if form == "fairness" and all_res.size > 1:
+                            # by default, log the std and decile
+                            new_results.pop(
+                                key, None)  # delete the redundant original one
+                            all_res.sort()
+                            new_results[f"{key}_std"] = np.std(
+                                np.array(all_res))
+                            new_results[f"{key}_bottom_decile"] = all_res[
+                                all_res.size // 10]
+                            new_results[f"{key}_top_decile"] = all_res[
+                                all_res.size * 9 // 10]
+                round_formatted_results[f'Results_{form}'] = new_results
+
+        with open(os.path.join(self.outdir, "eval_results.raw"),
+                  "a") as outfile:
+            outfile.write(str(round_formatted_results_raw) + "\n")
+
+        return round_formatted_results
+
+    def calc_blocal_dissim(last_model, local_updated_models):
+        '''
+        Arguments:
+            last_model (dict): the state of last round.
+            local_updated_models (list): each element is ooxx.
+        Returns:
+            b_local_dissimilarity (dict): the measurements.
+        '''
```

Comment:
please elaborate the metric of dissimilarity.

Reply:
Reference added